File: backend/coach/planCritic.ts
import { z } from "zod";

/**
 * Server-side plan critic — mirrors iOS `PlanCritic.swift` for worker validation after LLM plan generation.
 * Operates on decoded plan JSON before returning to the client.
 */

export const planCriticIssueSchema = z.object({
  code: z.string().max(64),
  message: z.string().max(500),
  severity: z.enum(["warning", "error"]),
});

export const planCriticVerdictSchema = z.object({
  issues: z.array(planCriticIssueSchema).max(32),
});

export type PlanCriticIssue = z.infer<typeof planCriticIssueSchema>;
export type PlanCriticVerdict = z.infer<typeof planCriticVerdictSchema>;

/** Minimal plan shape for critic checks (matches iOS TrainingPlan JSON). */
export const planCriticPlanSchema = z.object({
  weeks: z.array(
    z.object({
      weekNumber: z.number().int(),
      tssTarget: z
        .object({ lowerBound: z.number(), upperBound: z.number() })
        .optional(),
      tssTargetLower: z.number().optional(),
      tssTargetUpper: z.number().optional(),
      days: z.array(
        z.object({
          id: z.string(),
          weekNumber: z.number().int(),
          dayOfWeek: z.number().int().min(1).max(7),
          dayType: z.string(),
          title: z.string(),
          isKeyWorkout: z.boolean().optional(),
          durationMinutes: z.number().optional(),
          zone: z.string().optional(),
          intervals: z.array(z.unknown()).optional(),
        })
      ),
    })
  ),
});

export type PlanCriticPlanInput = z.infer<typeof planCriticPlanSchema>;

const MAX_WEEK_TO_WEEK_INCREASE = 0.15;
const MIN_REST_DAYS_PER_WEEK = 1;

const WORKOUT_TYPES = new Set([
  "workout",
  "ftpTest",
  "optionalWorkout",
  "commute",
]);

/** Lightweight week TSS estimate for critic (server may use richer math later). */
function estimateDayTSS(day: PlanCriticPlanInput["weeks"][number]["days"][number], ftp: number): number {
  if (!WORKOUT_TYPES.has(day.dayType)) return 0;
  const minutes = day.durationMinutes ?? 60;
  return (minutes * 60 * 0.65 * 0.65) / 36; // ~Z2 IF midpoint, same formula family as iOS
}

function weekTSS(week: PlanCriticPlanInput["weeks"][number], ftp: number): number {
  return week.days.reduce((sum, day) => sum + estimateDayTSS(day, ftp), 0);
}

/** Run critic rules aligned with iOS PlanCritic. */
export function validatePlanForCritic(plan: PlanCriticPlanInput, ftp = 250): PlanCriticVerdict {
  const issues: PlanCriticIssue[] = [];
  const weekly = plan.weeks.map((w) => weekTSS(w, ftp));

  for (let i = 1; i < weekly.length; i++) {
    const prev = weekly[i - 1]!;
    const curr = weekly[i]!;
    if (prev > 0 && (curr - prev) / prev > MAX_WEEK_TO_WEEK_INCREASE) {
      issues.push({
        code: "week_tss_spike",
        message: `Week ${i + 1} planned TSS jumps more than ${MAX_WEEK_TO_WEEK_INCREASE * 100}% vs week ${i}.`,
        severity: "warning",
      });
    }
  }

  for (const week of plan.weeks) {
    const workoutDays = week.days.filter((d) => WORKOUT_TYPES.has(d.dayType)).length;
    const restDays = Math.max(0, 7 - workoutDays);
    if (restDays < MIN_REST_DAYS_PER_WEEK) {
      issues.push({
        code: "insufficient_rest",
        message: `Week ${week.weekNumber} has fewer than ${MIN_REST_DAYS_PER_WEEK} rest/easy day(s).`,
        severity: "warning",
      });
    }

    const keyDays = week.days
      .filter((d) => d.isKeyWorkout && d.dayType !== "optionalWorkout" && d.dayType !== "rest")
      .sort((a, b) => a.dayOfWeek - b.dayOfWeek);

    for (let j = 1; j < keyDays.length; j++) {
      const prev = keyDays[j - 1]!;
      const curr = keyDays[j]!;
      if (curr.dayOfWeek - prev.dayOfWeek === 1) {
        issues.push({
          code: "back_to_back_key",
          message: `Week ${week.weekNumber}: key sessions "${prev.title}" and "${curr.title}" are on consecutive days.`,
          severity: "warning",
        });
      }
    }
  }

  return { issues };
}

export function criticWarningsFromVerdict(verdict: PlanCriticVerdict): string[] {
  return verdict.issues.filter((i) => i.severity === "warning").map((i) => i.message);
}

File: Mangox/Features/Training/Domain/UseCases/PlanCritic.swift
// Features/Training/Domain/UseCases/PlanCritic.swift
import Foundation

/// Client-side plan critic — validates proposed training plans before save (Phase 1).
enum PlanCritic {

    enum Severity: String, Sendable, Codable {
        case warning
        case error
    }

    struct Issue: Sendable, Equatable {
        let code: String
        let message: String
        let severity: Severity
    }

    struct Verdict: Sendable, Equatable {
        let issues: [Issue]
        var warnings: [Issue] { issues.filter { $0.severity == .warning } }
        var errors: [Issue] { issues.filter { $0.severity == .error } }
        var passed: Bool { errors.isEmpty }
    }

    static let maxWeekToWeekTSSIncrease = 0.15
    static let minRestDaysPerWeek = 1

    static func validate(plan: TrainingPlan, ftp: Int) -> Verdict {
        var issues: [Issue] = []
        let safeFTP = max(1, ftp)

        issues.append(contentsOf: weekToWeekLoadIssues(plan: plan, ftp: safeFTP))
        issues.append(contentsOf: restDayIssues(plan: plan))
        issues.append(contentsOf: keySessionSpacingIssues(plan: plan))
        issues.append(contentsOf: weeklyTargetAlignmentIssues(plan: plan, ftp: safeFTP))

        return Verdict(issues: issues)
    }

    private static func weekToWeekLoadIssues(plan: TrainingPlan, ftp: Int) -> [Issue] {
        let weeklyTSS = plan.weeks.map { week in
            week.days.reduce(0.0) { $0 + $1.estimatedPlannedTSS(ftp: ftp) }
        }
        guard weeklyTSS.count >= 2 else { return [] }

        var issues: [Issue] = []
        for index in 1..<weeklyTSS.count {
            let prev = weeklyTSS[index - 1]
            let curr = weeklyTSS[index]
            guard prev > 0 else { continue }
            let jump = (curr - prev) / prev
            if jump > maxWeekToWeekTSSIncrease {
                issues.append(
                    Issue(
                        code: "week_tss_spike",
                        message: String(
                            format: "Week %d planned TSS jumps %.0f%% vs week %d — keep increases under %.0f%%.",
                            index + 1,
                            jump * 100,
                            index,
                            maxWeekToWeekTSSIncrease * 100
                        ),
                        severity: .warning
                    )
                )
            }
        }
        return issues
    }

    private static func restDayIssues(plan: TrainingPlan) -> [Issue] {
        var issues: [Issue] = []
        for week in plan.weeks {
            let workoutDays = week.days.filter {
                switch $0.dayType {
                case .workout, .ftpTest, .optionalWorkout, .commute: return true
                default: return false
                }
            }.count
            let restDays = max(0, 7 - workoutDays)
            if restDays < minRestDaysPerWeek {
                issues.append(
                    Issue(
                        code: "insufficient_rest",
                        message: "Week \(week.weekNumber) has fewer than \(minRestDaysPerWeek) rest/easy day(s).",
                        severity: .warning
                    )
                )
            }
        }
        return issues
    }

    private static func keySessionSpacingIssues(plan: TrainingPlan) -> [Issue] {
        var issues: [Issue] = []
        for week in plan.weeks {
            let keyDays = week.days
                .filter { $0.isKeyWorkout && $0.dayType != .optionalWorkout && $0.dayType != .rest }
                .sorted { $0.dayOfWeek < $1.dayOfWeek }

            for pair in zip(keyDays, keyDays.dropFirst()) {
                if pair.1.dayOfWeek - pair.0.dayOfWeek == 1 {
                    issues.append(
                        Issue(
                            code: "back_to_back_key",
                            message: "Week \(week.weekNumber): key sessions \"\(pair.0.title)\" and \"\(pair.1.title)\" are on consecutive days.",
                            severity: .warning
                        )
                    )
                }
            }
        }
        return issues
    }

    private static func weeklyTargetAlignmentIssues(plan: TrainingPlan, ftp: Int) -> [Issue] {
        var issues: [Issue] = []
        for week in plan.weeks {
            let computed = week.days.reduce(0.0) { $0 + $1.estimatedPlannedTSS(ftp: ftp) }
            let targetMid = Double(week.tssTarget.lowerBound + week.tssTarget.upperBound) / 2
            guard targetMid > 0, computed > 0 else { continue }

            let deviation = abs(computed - targetMid) / targetMid
            if deviation > 0.25 {
                issues.append(
                    Issue(
                        code: "tss_target_mismatch",
                        message: String(
                            format: "Week %d computed TSS %.0f differs from target mid %.0f by %.0f%%.",
                            week.weekNumber,
                            computed,
                            targetMid,
                            deviation * 100
                        ),
                        severity: .warning
                    )
                )
            }
        }
        return issues
    }
}
